planetiler-core/src/main/java/com/onthegomap/planetiler/stream/WriteableProtoStreamArchive.java

```diff
@@ -0,0 +1,184 @@
+package com.onthegomap.planetiler.stream;
+
+import com.google.protobuf.ByteString;
+import com.onthegomap.planetiler.archive.TileArchiveMetadata;
+import com.onthegomap.planetiler.archive.TileEncodingResult;
+import com.onthegomap.planetiler.geo.TileCoord;
+import com.onthegomap.planetiler.proto.StreamArchiveProto;
+import com.onthegomap.planetiler.util.LayerStats.VectorLayer;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.UncheckedIOException;
+import java.nio.file.Path;
+import java.util.function.Consumer;
+import org.locationtech.jts.geom.CoordinateXY;
+import org.locationtech.jts.geom.Envelope;
+
+/**
+ * Writes protobuf-serialized tile data as well as meta data into file(s). The messages are of type
+ * {@link StreamArchiveProto.Entry} and are length-delimited.
+ */
+public final class WriteableProtoStreamArchive extends WritableStreamArchive {
```

Comment:
The proto output stream is neat, I'm just wondering how/when you were thinking we'd make use of it?

Reply:
Out of all formats the proto one should be the fastest to read/write given it's binary. If protobuf code generation is setup in a project, then it should also be the simplest to process given the model+parsing is generated then.
It should also be the least error-prone to use; parsing JSON even with jackson can be tricky if not done correctly: fail on unknown (i.e. new) properties, fail on unknown (i.e. new) enum values, ....

So for writing anything more serious/stables (let's say some pipeline that updates tiles in postgres - batched - with the tile ID or quadkey as primary key), I would probably choose the proto one over JSON.

I understand if you think that's too niche but given protobuf codegeneration is already setup, I think it's rather simple to add support for it. But other than it's faster+easier (depends case by case), most cases can be realized with JSON as well.

- CSV vs proto/JSON - probably boils down to whether meta data is needed
- proto vs CSV/JSON - probably boils down to binary vs text and "simplicity"/"speed" of parsing